downstairs/src/lib.rs

```diff
@@ -3220,139 +3221,125 @@ impl Work {
-    ///
-    /// If the dependencies are not met, move the state to `DepWait` and return
-    /// `false`.
+    /// Checks whether the given job is ready
     ///
-    /// If the job already is `InProgress`, return `true` (to be idempotent).
-    fn in_progress(&mut self, ds_id: JobId) -> bool {
-        let Some(job) = self.active.get_mut(&ds_id) else {
-            panic!("called in_progress for invalid job");
-        };
+    /// Updates `self.outstanding_deps` and `job.state`
+    fn check_ready(&mut self, job: &mut DownstairsWork) -> bool {
+        /*
+         * Before we can make this in_progress, we have to check the dep
+         * list if there is one and make sure all dependencies are
+         * completed.
+         */
+        let dep_list = job.work.deps();
 
-        match job.state {
-            WorkState::New | WorkState::DepWait => {
-                /*
-                 * Before we can make this in_progress, we have to check the dep
-                 * list if there is one and make sure all dependencies are
-                 * completed.
-                 */
-                let dep_list = job.work.deps();
+        /*
+         * See which of our dependencies are met.
+         * XXX Make this better/faster by removing the ones that
+         * are met, so next lap we don't have to check again?  There
+         * may be some debug value to knowing what the dep list was,
+         * so consider that before making this faster.
+         */
+        let mut deps_outstanding: Vec<JobId> =
+            Vec::with_capacity(dep_list.len());
 
-                /*
-                 * See which of our dependencies are met.
-                 * XXX Make this better/faster by removing the ones that
-                 * are met, so next lap we don't have to check again?  There
-                 * may be some debug value to knowing what the dep list was,
-                 * so consider that before making this faster.
-                 */
-                let mut deps_outstanding: Vec<JobId> =
-                    Vec::with_capacity(dep_list.len());
-
-                for dep in dep_list.iter() {
-                    // The Downstairs currently assumes that all jobs previous
-                    // to the last flush have completed, hence this early out.
-                    //
-                    // Currently `work.completed` is cleared out when
-                    // `Downstairs::complete_work` (or `complete` in mod test)
-                    // is called with a FlushAck so this early out cannot be
-                    // removed unless that is changed too.
-                    if dep <= &self.last_flush {
-                        continue;
-                    }
+        for dep in dep_list.iter() {
+            // The Downstairs currently assumes that all jobs previous
+            // to the last flush have completed, hence this early out.
+            //
+            // Currently `work.completed` is cleared out when
```

Comment:
Not sure when `Downstairs::complete_work` went away, but it's not there now.
Non test jobs updates `work.completed` in `do_ready_work()`, so we should replace
`Downstairs::complete_work` with that.

Reply:
Also fixed in 9587ae9b547f811f862236b5793174fa87e2d594